processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -542,46 +327,299 @@ protected T _get(ByteBuffer copyValueBuffer, int startOffset, int endOffset)
+  {
+    boolean allowReverseLookup = byteBuffer.get() == REVERSE_LOOKUP_ALLOWED;
+    int size = byteBuffer.getInt();
+    ByteBuffer bufferToUse = byteBuffer.asReadOnlyBuffer();
+    bufferToUse.limit(bufferToUse.position() + size);
+    byteBuffer.position(bufferToUse.limit());
 
-      value = (value != null && value.equals("")) ? null : value;
+    return new GenericIndexed<>(
+        bufferToUse,
+        strategy,
+        allowReverseLookup
+    );
+  }
 
-      int minIndex = 0;
-      int maxIndex = size - 1;
-      while (minIndex <= maxIndex) {
-        int currIndex = (minIndex + maxIndex) >>> 1;
+  private static <T> GenericIndexed<T> fromIterableVersionOne(Iterable<T> objectsIterable, ObjectStrategy<T> strategy)
+  {
+    Iterator<T> objects = objectsIterable.iterator();
+    if (!objects.hasNext()) {
+      final ByteBuffer buffer = ByteBuffer.allocate(Ints.BYTES).putInt(0);
+      buffer.flip();
+      return new GenericIndexed<>(buffer, strategy, true);
+    }
 
-        T currValue = get(currIndex);
-        int comparison = strategy.compare(currValue, value);
-        if (comparison == 0) {
-          return currIndex;
+    boolean allowReverseLookup = true;
+    int count = 0;
+
+    ZeroCopyByteArrayOutputStream headerBytes = new ZeroCopyByteArrayOutputStream();
+    ZeroCopyByteArrayOutputStream valueBytes = new ZeroCopyByteArrayOutputStream();
+    ByteBuffer helperBuffer = ByteBuffer.allocate(Ints.BYTES);
+    try {
+      int offset = 0;
+      T prevVal = null;
+      do {
+        count++;
+        T next = objects.next();
+        if (allowReverseLookup && prevVal != null && !(strategy.compare(prevVal, next) < 0)) {
+          allowReverseLookup = false;
         }
 
-        if (comparison < 0) {
-          minIndex = currIndex + 1;
-        } else {
-          maxIndex = currIndex - 1;
+        final byte[] bytes = strategy.toBytes(next);
+        offset += Ints.BYTES + bytes.length;
+        GenericIndexedWriter.writeIntValueToOutputStream(helperBuffer, offset, headerBytes);
+        GenericIndexedWriter.writeIntValueToOutputStream(helperBuffer, bytes.length, valueBytes);
+        valueBytes.write(bytes);
+
+        if (prevVal instanceof Closeable) {
+          CloseQuietly.close((Closeable) prevVal);
         }
+        prevVal = next;
+      } while (objects.hasNext());
+
+      if (prevVal instanceof Closeable) {
+        CloseQuietly.close((Closeable) prevVal);
       }
+    }
+    catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    ByteBuffer theBuffer = ByteBuffer.allocate(Ints.BYTES + headerBytes.size() + valueBytes.size());
+    theBuffer.putInt(count);
+    headerBytes.writeTo(theBuffer);
+    valueBytes.writeTo(theBuffer);
+    theBuffer.flip();
+
+    return new GenericIndexed<>(theBuffer.asReadOnlyBuffer(), strategy, allowReverseLookup);
+  }
+
+  GenericIndexed(
```

Comment:
I generally expect the constructors to come after the fields.  Threw me for a loop initially trying to read this code wondering if you had made it abstract and expected the children to modify fields directly, but no, the fields are `private` so that can't be.  Then I noticed that you have the constructor stuck way down here under the actual instance methods.  There's a general ordering to methods and fields that is followed in the rest of the Druid codebase, please stay consistent with that.